ir/deserializer.go

```diff
@@ -0,0 +1,284 @@
+			if tsInt, err := strconv.ParseInt(tsStr, 10, 64); nil == err {
+				refTs = ffi.EpochTimeMs(tsInt)
+				*(*ffi.EpochTimeMs)(timestampCptr) = refTs
+			}
+		}
+		deserializer = &fourByteDeserializer{commonDeserializer{tsInfo, deserializerCptr}, refTs}
+	} else {
+		deserializer = &eightByteDeserializer{commonDeserializer{tsInfo, deserializerCptr}}
+	}
+
+	return deserializer, int(pos), nil
+}
+
+// commonDeserializer contains fields common to all types of CLP IR encoding.
+// TimestampInfo stores information common to all timestamps found in the IR.
+// cptr holds a reference to the underlying C++ objected used as backing storage
+// for the Views returned by the deserializer. Close must be called to free this
+// underlying memory and failure to do so will result in a memory leak.
+type commonDeserializer struct {
+	tsInfo TimestampInfo
+	cptr   unsafe.Pointer
+}
+
+// Close will delete the underlying C++ allocated memory used by the
+// deserializer. Failure to call Close will result in a memory leak.
+func (self *commonDeserializer) Close() error {
+	if nil != self.cptr {
+		C.ir_deserializer_close(self.cptr)
+		self.cptr = nil
+	}
+	return nil
+}
+
+// Returns the TimestampInfo used by the Deserializer.
+func (self commonDeserializer) TimestampInfo() TimestampInfo {
+	return self.tsInfo
+}
+
+type eightByteDeserializer struct {
+	commonDeserializer
+}
+
+// DeserializeLogEvent attempts to read the next log event from the IR stream in
+// irBuf, returning the deserialized [ffi.LogEventView], the position read to in
+// irBuf (the end of the log event in irBuf), and an error. On error returns:
+//   - nil *ffi.LogEventView
+//   - 0 position
+//   - [IrError] error: CLP failed to successfully deserialize
+//   - [EndOfIr] error: CLP found the IR stream EOF tag
+func (self *eightByteDeserializer) DeserializeLogEvent(
+	irBuf []byte,
+) (*ffi.LogEventView, int, error) {
+	return deserializeLogEvent(self, irBuf)
+}
+
+// fourByteDeserializer contains both a common CLP IR deserializer and stores
+// the previously seen log event's timestamp. The previous timestamp is
+// necessary to calculate the current timestamp as four byte encoding only
+// encodes the timestamp delta between the current log event and the previous.
+type fourByteDeserializer struct {
+	commonDeserializer
+	prevTimestamp ffi.EpochTimeMs
+}
+
+// DeserializeLogEvent attempts to read the next log event from the IR stream in
+// irBuf, returning the deserialized [ffi.LogEventView], the position read to in
+// irBuf (the end of the log event in irBuf), and an error. On error returns:
+//   - nil *ffi.LogEventView
+//   - 0 position
+//   - [IrError] error: CLP failed to successfully deserialize
+//   - [EndOfIr] error: CLP found the IR stream EOF tag
+func (self *fourByteDeserializer) DeserializeLogEvent(
+	irBuf []byte,
+) (*ffi.LogEventView, int, error) {
+	return deserializeLogEvent(self, irBuf)
+}
+
+func (self *eightByteDeserializer) DeserializeWildcardMatch(
```

Comment:
maybe we should move this function following `(self *eightByteDeserializer) DeserializieLogEvent`